### PR TITLE
Batch messages sent over DDP from server to client (DDP V2?)

### PR DESCRIPTION
**Pull request by KoenLav**
Messages sent from the client to the server were already buffered, but still JSON.parse would be called once for every message and while buffering was implemented for messages from the client to the server, it was not implemented for messages from server to client.

This implementation:
- sends the buffered messages from client to server as an array;
- implements the same buffering strategy on the server side;
- enables both the client and the server to interpret batched DDP messages.

Requesting feedback!

ps this PR supersedes: https://github.com/meteor/meteor/pull/9855
pps I'll leave some considerations as comments in the code.

### CODE REVIEW

File: packages/ddp-client/common/livedata_connection.js [hunk 1]
```diff
@@ -1645,47 +1668,54 @@ export class Connection {
...
-    if (msg === null || !msg.msg) {
-      // XXX COMPAT WITH 0.6.6. ignore the old welcome message for back
-      // compat.  Remove this 'if' once the server stops sending welcome
-      // messages (stream_server.js).
-      if (!(msg && msg.server_id))
-        Meteor._debug('discarding invalid livedata message', msg);
-      return;
-    }
+    // Convert all DDP messages to Array form
+    messages = messages.constructor === Array ? messages : [messages];
```
[the remark this answers is not in the document]
Comment (sebakerckhof) [reply packages/ddp-client/common/livedata_connection.js]: Array.isArray ?

**Comment by oswaldoacauan** [comment]
I would gladly help to merge this PR, need this feature asap :)

**Comment by KoenLav** [comment]
@oswaldoacauan developing for Meteor is rather simple; just fork the repository and follow the instructions in DEVELOPMENT.MD.

The guys from Meteor indicated they will focus on releasing 1.7 first and then work towards resolutions for server performance issues (like this PR is supposed to address).

Are there any insights you could share as to the specific problem you are experiencing? Or did you already share them?

**Comment by mitar** [comment]
Hm, can you show how this improves performance? I do not really understand why there would be much performance change. Can you make an example app where this shows?

**Comment by KoenLav** [comment]
@mitar in a scenario where a server would want to send 500 added/changed/removed messages within 500ms currently these messages are:
- passed to JSON.stringify individually;
- sent to the client individually.

Both of these operations have overhead associated with them, which is now only incurred once (at the cost of stringifying and sending a larger amount of data).

I expect this change to have a large impact on the following issues:
- https://github.com/meteor/meteor/issues/9828
- https://github.com/meteor/meteor/issues/9796

Reproduction: https://github.com/KoenLav/meteor-fiber-repro

Would love to see a test showing the performance improvement as well, but no time to create one currently.

File: packages/ddp-common/utils.js [hunk 2]
```diff
@@ -42,76 +42,94 @@ DDPCommon.SUPPORTED_DDP_VERSIONS = [ '1', 'pre2', 'pre1' ];
...
+  const clonedMessages = [];
 
-  // swizzle 'changed' messages from 'fields undefined' rep to 'fields
-  // and cleared' rep
-  if (hasOwn.call(msg, 'fields')) {
-    const cleared = [];
+  keys(messages).forEach(i => {
+    const msg = messages[i];
+    
+    const copy = EJSON.clone(msg);
```
[the remark this answers is not in the document]
Comment (sebakerckhof) [reply packages/ddp-common/utils.js]: as @copleykj also pointed out. There's no need for `keys`. Just do `messages.forEach(msg => ...`

**Comment by crapthings** [comment]
@KoenLav will this add new ddp message type or use exist one ?

does this change DDP Specification?

**Comment by KoenLav** [comment]
@crapthings 

No, it simply batches the messages.

Yes, it does change the specification as clients will need to know how to accept arrays of messages rather than singular messages, however the intention is to make this fully backwards compatible for 'older' clients AND/OR an optional feature.

**Comment by mitar** [comment]
> I expect this change to have a large impact on the following issues:

But you should probably confirm this. It might be an unnecessary change if underlying fiber pool issue is fixed. I think it might be better approach to get node used by Meteor to have a fix addressing the fiber pool issue. In that case both this issue and any other code path which makes a spike of fibers would perform better. This is why we need a test to measure performance improvement. So that we could compare existing code vs. existing code with fiber patch vs. existing code with batching of messages vs. existing code with fiber patch with batching of messages. Then we can know which approach to take.

**Comment by crapthings** [comment]
serializing 20000 items individually vs serializing once

=> Meteor server restarted
I20180509-09:27:48.820(8)? foreach: 94.341ms
=> Meteor server restarted
I20180509-09:28:30.809(8)? a: 104.490ms
I20180509-09:28:30.903(8)? b: 92.962ms
=> Meteor server restarted
I20180509-09:28:43.452(8)? a: 91.389ms
I20180509-09:28:43.566(8)? b: 113.280ms
=> Meteor server restarted
I20180509-09:28:48.563(8)? a: 102.033ms
I20180509-09:28:48.663(8)? b: 99.924ms
=> Meteor server restarted
I20180509-09:28:54.795(8)? a: 101.871ms
I20180509-09:28:54.902(8)? b: 106.864ms
=> Meteor server restarted
I20180509-09:29:04.691(8)? a: 103.247ms
I20180509-09:29:04.793(8)? b: 101.476ms
=> Meteor server restarted

```js
import _ from 'lodash'
import faker from 'faker'

const list = _.times(20000, n => ({
  title: faker.lorem.sentence(),
  text: faker.lorem.paragraphs(),
  category: faker.lorem.word(),
  tags: (faker.lorem.words()).split(' '),
}))

console.time('a')
const a = _.map(list, item => JSON.stringify(item))
console.timeEnd('a')

console.time('b')
const b = JSON.stringify(list)
console.timeEnd('b')
```



METEOR@1.6.1.1

### publish this.added

2w items one by one
image

2w items in one
image

2w items in 4 messages
image

2w items in 50 messages
image

so we need batching size control ?

i think 'this.removed' is similar to 'this.added', remove just send id.

**Comment by mitar** [comment]
> serializing 20000 items individually vs serializing once

Your results literally show that it does not matter. Once one is faster, another time the other is faster. You should probably repeat this 10 times and average. But from the look of it, it really looks like there is not much to gain here, except new complexity.

> but cpu spike is in fiber

Yes, I think we should fix fiber spiking with a patch in node. But I am not yet convinced about batching at the API level. How did you measure things on the client side? So where is code for those outputs in the client console you made screen shots of? So processing each message separately on the client is know to be slow. Client has many things to optimize.

File: packages/ddp-common/utils.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (copleykj) [reply packages/ddp-common/utils.js]: Also being that the point of this PR is performance, I would recommend a regular for loop here as it's ~90% faster

**Comment by copleykj** [comment]
There does seem to be at least some merit in this effort as far as data being sent in small batches.

**Comment by crapthings** [comment]
@mitar 

```
Meteor.startup(() => {
  console.time('testa')
  Tracker.autorun(() => {
    const ready = Meteor.subscribe('testc').ready()
    if (!ready) return
    console.log(Test.find().fetch())
    console.timeEnd('testa')
  })
})
```

**Comment by mitar** [comment]
That looks OK for testing, but still it would be good to know where exactly is slow because of lack of batching. It might be that only client-side batching is necessary.

> and when u stop subscription, the data in cursor removed one by one, if its in batching it will be fast.

Yea, I can imagine that those operations on client are slow when removing one by one. But this could be fixed on client side only.

**Comment by KoenLav** [comment]
@mitar the problem is that some of these functions on the server side are blocking and that when a lot of these blocking events occur at the same time the server remains stuck at 100% CPU.

The patch for V8 which limits the impact of this issue is already in the works, however I believe we can also make some smart changes to Meteor which help in actually resolving the cause of the issue, rather than minimizing the effect.

In my current PR the messages are batched within a particular Session instance (so per connected user). Suppose there are 50 connected users and 2000 changes within half a second this means 100.000 stacks get executed, at once.

By batching the messages in the Session, the code leading up from the Collection observer up until the Stringify function is still called 100.000 times, but the code afterwards (stringifying and sending the messages) only 50 times.

I'm thinking about the possibility of batching messages, per Collection, as Meteor receives them from MongoDB. This would significantly reduce the load on the server as all of the code leading up to the sending of DDP messages would only be executed 50 times, rather than 100.000 times, and the code which batches the messages would only be executed 2000 times.

This would produce 2000 short stacks, batching the messages, eventually leading to 50 long stacks (with a bigger payload) instead of 100.000 long stacks with small payloads.

**Comment by tdjac0bs** [comment]
We was going to implement the same improvement to Meteor, because we figured out that the performance is affected due to the chatty nature of the DDP protocol (specially when we have a high RTT). Batching the messages server side solves this chatty problem.

@oswaldoacauan, can you create a demo app that make 10.000 updates per second, during 120s and compare the performance? 

I suggest the following tests:
- Run the demo app locally (latency 0) without this PR.
- Run the demo app locally (latency 0) with this PR.
- Run the demo app locally (latency 150ms) without this PR.
- Run the demo app locally (latency  150ms) with this PR.

The measured variable could be the time the client takes to get all the updates.

**Comment by KoenLav** [comment]
I've just started working on batching the messages in the ObserveMutiplexer; the first place where the changes in a subscription are handled.

This will require refactoring every class involved (Session, Subscription, Observer, etc) but will completely solve the issues. 

Will post WIP soon.

**Comment by sebakerckhof** [comment]
> This will require refactoring every class involved (Session, Subscription, Observer, etc) but will completely solve the issues.

Before refactoring such core classes, it's really recommended to open a feature request and have a discussion in there.

**Comment by KoenLav** [comment]
@mitar 

ObserveMultiplexer._applyCallback creates a SynchronousQueue entry for every change to a Collection which we are Subscribed to OR Observing (Subscriptions make use of Observe, which makes use of ObserveChanges, internally).

We would not want to handle these changes asynchronously, so this seems fine, but the way Meteor is structured the entire callstack

@sebakerckhof

In my opinion the easiest way to get some discussion started it by having a suggested implementation. I do agree refactoring these classes requires significant discussion and improvements to the current situation.

Then again, if we would implement the changes in a backwards-compatible (opt-in) way and are able to show significant performance improvements (and or the resolution of several long-standing issues) I think we should fix first, discuss and improve later.

**Comment by KoenLav** [comment]
Added some commits; the implementation is now fully backward compatible. In addition to that I ran some tests on the Fiber spikes repository (issue: https://github.com/meteor/meteor/issues/7747 , repository: https://github.com/KoenLav/meteor-fiber-repro) and can confirm that this PR improves the performance considerably. I would welcome some more tests/use cases of this specific batching implementation!

Even though performance increased considerably and time Meteor maxes out CPU is significantly reduced, the Fiber spike remains.

In an effort to find the source of the spike I started refactoring the aforementioned classes (Multiplexer, Observer, etc.), but I have not been successful in locating the source yet. In this PR, however, batching is applied at two levels, please take a look: https://github.com/meteor/meteor/pull/9876

**Comment by KoenLav** [comment]
@mitar (and others) I tested the change in throughput caused by these changes on my development machine; loading the 20.000 documents (Widgets) in the Meteor Fibers Repro repository takes 20-40 seconds without these changes and 1-2 seconds with these changes.

It's fairly easy to test and see for yourself!

I no longer consider this PR to be a WIP, from my side, so would like to get some more comments :)

**Comment by mitar** [comment]
Can you also test this running node version with a patch by Kenton which fixes the underlying fiber issue?

**Comment by mitar** [comment]
> @mitar this patch improves performance with the aforementioned numbers even before the Fibers issue is triggered (when simply first loading a page which subscribes to 20.000 documents). So it should be considered an improvement regardless.

Why? If that other patch maybe fixes both this and any other fiber spiking issue? I think this is great work, but we should really understand if it is needed on its own or it is just a workaround for a fiber spiking issue.

**Comment by KoenLav** [comment]
The Fibers which get created are all created (and exited) within the OplogObserveDriver, they actually don't have anything to do with the changes in this PR, other than that they are also involved in the process of moving oplog notifications from MongoDB through the server all the way to the client.

Maybe I'm not providing enough details, but I do now in fact understand the various problems and know what their fixes are and how they are interrelated.

If I can help clarify any specific questions I'd be happy to do so!

**Comment by kentonv** [comment]
To be clear, my change in V8 does not in any way change the number of fibers an app creates, it only prevents V8 from having pathologically bad performance when the fiber count gets high. It's still definitely a good idea to avoid high fiber counts, as each fiber allocates a stack and other resources.

**Comment by nathan-muir** [comment]
I've been following this PR as I'm quite interested in the fiber-spike issue.

I agree with @mitar that It would be good to compare this patch, with something that addresses the fiber spikes directly.

So [here's my attempt](https://github.com/meteor/meteor/compare/release-1.7...nathan-muir:nm-oplog-fiber-pool), with a [slightly updated reproduction](https://github.com/nathan-muir/meteor-fiber-repro) (Adding timing and using meteor-1.7rc.8)

With vanilla Meteor-1.7, I'm seeing 20000+ fiber created, and about 10s to `unset` and 20s - 30s to `set` on the client.

image

With my patch applied to the `DocFetcher`, I'm seeing 45 fiber created, and the same `10s` to unset, but only 6s to `set` on the client.

image

**Comment by KoenLav** [comment]
In my opinion placing the Fibers in a pool is not resolving the issue, rather masquerading the effects.

Also, this PR does NOT address the Fibers issue in any way, it's simply a way of communicating with clients more efficiently.

Only my latest PR, which actually fixes the Fiber issue, addresses that one, all the other PRs are just general improvements to Meteor performance.

**Comment by KoenLav** [comment]
Adding to that; when Fiber spikes do arise adding them to a pool makes a lot of sense, so this might be a good fallback for Meteor to implement either way.

**Comment by KoenLav** [comment]
@nathan-muir you are correct in saying that this patch does literally nothing to alleviate the creation of Fibers on the server (as these are created in the DocFetcher and this PR does not touch any code 'below' that level (not even close).

I would be very interested in seeing (1) your patch to DocFetcher and (2) the performance of an application where that patch is applied together with this PR (it batches the messages right after the (Oplog)ObserveDriver): https://github.com/meteor/meteor/pull/9876

**Comment by nathan-muir** [comment]
@KoenLav Here's the patch https://github.com/meteor/meteor/compare/release-1.7...nathan-muir:nm-oplog-fiber-pool#diff-6262b939ca8b8ddd9f3013457526cd05


You can cherry pick it from my fork like so:

```
git remote add tmp git@github.com:nathan-muir/meteor.git
git fetch nathan-muir
git cherry-pick 6fc0d83 5ee8eb5 906c7bb
git remote remove tmp
```

**Comment by KoenLav** [comment]
@nathan-muir I see... Could you explain why using a pool for the Fibers makes such a performance issue?

Im addition to that I still feel that, when there are multiple docs to be fetched, it would be better to fetch multiple docs by id (as in create one Fiber which fetches a thousand docs and processes them, rather than a thousand Fibers fetching one doc each).

What do you think about this?

**Comment by KoenLav** [comment]
This PR has been rebased against 1.7.0.3.

To review and/or merge I suggest merging modernize-ddp-server (https://github.com/meteor/meteor/pull/9622) first.

**Comment by KoenLav** [comment]
@abernix @benjamn

**Comment by StorytellerCZ** [comment]
As stated in #9885 , always merge PRs against devel branch.
@benjamn @abernix

**Pull request closed without merging**